### PR TITLE
updating v0.3 to use substrate v0.10

### LINKED ISSUE #133: Update v0.3 to latest Substrate v0.10
polkadot v0.3 -> substrate v0.10

### PR DESCRIPTION
**Pull request by gterzian**
Fix #133

### CODE REVIEW

**Comment by gterzian** [comment]
```
 --> service/src/lib.rs:150:1
    |
150 | / construct_service_factory! {
151 | |     struct Factory {
152 | |         Block = Block,
153 | |         RuntimeApi = RuntimeApi,
...   |
275 | |     }
276 | | }
    | |_^ the trait `sr_primitives::traits::ProvideRuntimeApi` is not implemented for `grandpa::import::GrandpaBlockImport<substrate_client_db::Backend<sr_primitives::generic::Block<sr_primitives::generic::Header<u64, sr_primitives::traits::BlakeTwo256, sr_primitives::generic::DigestItem<primitives::H256, primitives::Ed25519AuthorityId>>, polkadot_primitives::UncheckedExtrinsic>>, client::LocalCallExecutor<substrate_client_db::Backend<sr_primitives::generic::Block<sr_primitives::generic::Header<u64, sr_primitives::traits::BlakeTwo256, sr_primitives::generic::DigestItem<primitives::H256, primitives::Ed25519AuthorityId>>, polkadot_primitives::UncheckedExtrinsic>>, substrate_executor::native_executor::NativeExecutor<polkadot_executor::Executor>>, sr_primitives::generic::Block<sr_primitives::generic::Header<u64, sr_primitives::traits::BlakeTwo256, sr_primitives::generic::DigestItem<primitives::H256, primitives::Ed25519AuthorityId>>, polkadot_primitives::UncheckedExtrinsic>, polkadot_runtime::RuntimeApi, client::Client<substrate_client_db::Backend<sr_primitives::generic::Block<sr_primitives::generic::Header<u64, sr_primitives::traits::BlakeTwo256, sr_primitives::generic::DigestItem<primitives::H256, primitives::Ed25519AuthorityId>>, polkadot_primitives::UncheckedExtrinsic>>, client::LocalCallExecutor<substrate_client_db::Backend<sr_primitives::generic::Block<sr_primitives::generic::Header<u64, sr_primitives::traits::BlakeTwo256, sr_primitives::generic::DigestItem<primitives::H256, primitives::Ed25519AuthorityId>>, polkadot_primitives::UncheckedExtrinsic>>, substrate_executor::native_executor::NativeExecutor<polkadot_executor::Executor>>, sr_primitives::generic::Block<sr_primitives::generic::Header<u64, sr_primitives::traits::BlakeTwo256, sr_primitives::generic::DigestItem<primitives::H256, primitives::Ed25519AuthorityId>>, polkadot_primitives::UncheckedExtrinsic>, polkadot_runtime::RuntimeApi>>`
    |
    = note: required because of the requirements on the impl of `substrate_consensus_common::import_queue::Verifier
```

Anyone can make-out what this entails?

**Comment by rphmeier** [comment]
We used to forward these traits on `GrandpaBlockImport` but it seems we don't anymore. Maybe we need to do that again upstream -- @sorpaas removed them because they weren't being used.

**Comment by gterzian** [comment]
@bkchr Would you have any pointers with regards to updating polkadot in the light of https://github.com/paritytech/substrate/commit/53bf81e57cdcae34f50bb9359813053e9498b1cd#diff-2d4fe59416e64524ebe5d49fb1cd82c2

I think it would be mostly about implementing [`ProvideInherent`](https://github.com/paritytech/polkadot/pull/146/files#diff-a0b66d4bafdfbafd411b7384384df493L423)

as well as [`propose`](https://github.com/paritytech/polkadot/pull/146/files#diff-7ebbb9bce1cefe0f24c54625cca34828R521) and [`propose_with`](https://github.com/paritytech/polkadot/pull/146/files#diff-7ebbb9bce1cefe0f24c54625cca34828R650)

**Comment by bkchr** [comment]
Polkadot is already using the new inherent data interface in master.

File: consensus/src/lib.rs [hunk 1]
```diff
@@ -505,20 +514,21 @@ pub struct Proposer<C: Send + Sync, TxApi: PoolChainApi> where
...
+	C::Api: ParachainHost<Block> + BlockBuilderApi<Block>,
 {
 	type Error = Error;
 	type Create = Either<
 		CreateProposal<C, TxApi>,
 		future::FutureResult<Block, Error>,
 	>;
 
-	fn propose(&self, consensus_data: AuraConsensusData) -> Self::Create {
+	fn propose(&self, inherent_data: InherentData, max_duration: Duration) -> Self::Create {
```
Comment (rphmeier) [inline consensus/src/lib.rs]: @bkchr can you review this function to make sure the inherent data is backwards compatible with before?

File: runtime/Cargo.toml [hunk 2]
```diff
@@ -49,6 +50,7 @@ std = [
 	"polkadot-primitives/std",
 	"parity-codec/std",
 	"parity-codec-derive/std",
+    "substrate-inherents/std",
```
Comment (rphmeier) [inline runtime/Cargo.toml]: funky indentation

File: network/src/consensus.rs [hunk 3]
```diff
@@ -152,24 +152,35 @@ impl<P, E> Network for ConsensusNetwork<P,E> where
...
+		self.network.with_gossip(move |gossip, _| {
+			let inner_rx = gossip.messages_for(attestation_topic);
+			let _ = tx.send(inner_rx);
+		});
+
+		let table_router_clone = table_router.clone();
+		let executor = task_executor.clone();
+
+		// spin up a task in the background that processes all incoming statements
+		// TODO: propagate statements on a timer?
```
Comment (rphmeier) [inline network/src/consensus.rs]: this seems in duplicate with the above comment.

File: network/src/consensus.rs [hunk 4]
```diff
@@ -152,24 +152,35 @@ impl<P, E> Network for ConsensusNetwork<P,E> where
...
+		self.network
+			.with_spec(move |spec, ctx| {
```
Comment (rphmeier) [inline network/src/consensus.rs]: this whole block could be moved up a level of indentation if we didn't do a newline on `.with_spec`

File: network/src/consensus.rs [hunk 5]
```diff
@@ -180,14 +191,30 @@ impl<P, E> Network for ConsensusNetwork<P,E> where
...
+		if let Some(ref mut inner) = self.inner {
+			return inner
+				.poll()
+				.map_err(|_| NetworkDown)
+		}
+		match self.outer.poll() {
+			Ok(futures::Async::Ready(mut inner)) => {
+				let poll_result = inner.poll();
+				self.inner = Some(inner);
+				poll_result.map_err(|_| NetworkDown)
```
Comment (rphmeier) [inline network/src/consensus.rs]: seems more resilient to `return self.poll()` here.

File: network/src/consensus.rs [hunk 6]
```diff
@@ -180,14 +191,30 @@ impl<P, E> Network for ConsensusNetwork<P,E> where
 pub struct NetworkDown;
 
 /// A future that resolves when a collation is received.
-pub struct AwaitingCollation(::futures::sync::oneshot::Receiver<Collation>);
+pub struct AwaitingCollation {
+	outer: ::futures::sync::oneshot::Receiver<::futures::sync::oneshot::Receiver<Collation>>,
```
Comment (rphmeier) [inline network/src/consensus.rs]: I would recommend bringing `Receiver` into scope.

File: network/src/consensus.rs [hunk 7]
```diff
@@ -198,13 +225,16 @@ impl<P: ProvideRuntimeApi + Send + Sync + 'static, E: Clone> Collators for Conse
 	type Collation = AwaitingCollation;
 
 	fn collate(&self, parachain: ParaId, relay_parent: Hash) -> Self::Collation {
-		AwaitingCollation(
-			self.network.with_spec(|spec, _| spec.await_collation(relay_parent, parachain))
-		)
+		let (tx, rx) = ::futures::sync::oneshot::channel();
```
Comment (rphmeier) [inline network/src/consensus.rs]: perhaps oneshot being in scope is reasonable.

File: network/src/consensus.rs [hunk 8]
```diff
@@ -198,13 +225,16 @@ impl<P: ProvideRuntimeApi + Send + Sync + 'static, E: Clone> Collators for Conse
...
+		self.network.with_spec(move |spec, _| {
+			let collation = spec.await_collation(relay_parent, parachain);
+			let _ = tx.send(collation);
+		});
+		AwaitingCollation{outer: rx, inner: None}
```
Comment (rphmeier) [inline network/src/consensus.rs]: style: `AwaitingCollation { ... }` (note the spaces)

File: network/src/consensus.rs (hunk 3 above)
Comment (gnunicorn) [reply network/src/consensus.rs]: (Also, "no commit of TODO/FIXME without a Link" - I'd vote for removing that ToDo-comment at all)

File: primitives/Cargo.toml [hunk 9]
```diff
@@ -6,16 +6,16 @@ authors = ["Parity Technologies <admin@parity.io>"]
...
 serde_derive = { version = "1.0", optional = true }
-parity-codec = { version = "2.1", default-features = false }
-parity-codec-derive = { version = "2.1", default-features = false }
-substrate-primitives = { git = "https://github.com/paritytech/substrate", branch = "alexander-backports", default-features = false }
-substrate-client = { git = "https://github.com/paritytech/substrate", branch = "alexander-backports", default-features = false }
-sr-version = { git = "https://github.com/paritytech/substrate", branch = "alexander-backports", default-features = false }
-sr-std = { git = "https://github.com/paritytech/substrate", branch = "alexander-backports", default-features = false }
-sr-primitives = { git = "https://github.com/paritytech/substrate", branch = "alexander-backports", default-features = false }
+parity-codec = "3.0"
+parity-codec-derive = "3.0"
```
Comment (gnunicorn) [inline primitives/Cargo.toml]:
```suggestion
parity-codec-derive = { version = "3.0", default-features = false}
```

File: runtime/wasm/Cargo.toml [hunk 10]
```diff
@@ -13,29 +13,29 @@ polkadot-primitives = { path = "../../primitives", default-features = false }
 safe-mix = { version = "1.0", default-features = false }
 parity-codec = { version = "2.1", default-features = false }
```
Comment (gnunicorn) [inline runtime/wasm/Cargo.toml]: why keep the old version here?
```suggestion
 parity-codec = { version = "3.0", default-features = false }
```

File: runtime/wasm/Cargo.toml [hunk 11]
```diff
@@ -13,29 +13,29 @@ polkadot-primitives = { path = "../../primitives", default-features = false }
...
 parity-codec-derive = { version = "2.1", default-features = false }
```
Comment (gnunicorn) [inline runtime/wasm/Cargo.toml]:
```suggestion
 parity-codec-derive = { version = "3.0", default-features = false }
```

File: test-parachains/adder/collator/src/main.rs [hunk 12]
```diff
@@ -137,6 +137,7 @@ fn main() {
 			executable_name: "adder-collator",
 			description: "collator for adder parachain",
 			author: "parity technologies",
+			support_url: "none"
```
Comment (gnunicorn) [inline test-parachains/adder/collator/src/main.rs]: shouldn't collator issues still be reported to us?

```suggestion
			support_url: "https://github.com/paritytech/polkadot/issues/new",
```

File: consensus/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply consensus/src/lib.rs]: Do we need to produce blocks with the new node and the old runtime? If yes, then this is not compatible. For syncing, this code should not be touched.

**Comment by gterzian** [comment]
@bkchr  How can we address https://github.com/paritytech/polkadot/pull/146/files#r259722390 ?

File: consensus/src/lib.rs (hunk 1 above)
Comment (rphmeier) [reply consensus/src/lib.rs]: Ok, thanks for the clarification. I think we can work around this. What would happen if the old code tried to author on the new runtime?

Comment (bkchr) [reply consensus/src/lib.rs]: Create inherent will fail and thus the block production will fail.

**Pull request merged**